## 4.8 Grouping data and aggregating variables

## This script contains the following scripts:

1. Import libraries and data

2. Find the aggregated mean of the "number_of_order" column grouped by "department_id" for the subset and the entire dataframe

3. Analyze the results and differences

4. Create a loyalty flag using transform() and loc()

5. Check the basic statistics of the product prices for each loyalty category

6. Create a spending flag for each user based on the average prices across all their orders

7. Create an order frequency flag

8. Ensure the notebook is clean and structured and that the code is well commented

9. Export the dataframe as a pickle file in Prepared Data folder

## 1. Import libraries and data

In [6]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [7]:
# Import data

path = r'C:\Users\admin\06-2024 Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_new_cols.pkl'))

## 2. Find the aggregated mean of the "number_of_order" column grouped by "department_id" for the subset and the entire dataframe

In [9]:
# Create a subset

df = ords_prods_merge[:1000000]

In [10]:
df.shape

(1000000, 23)

In [11]:
df.head(10)

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   
5             0           1  Chocolate Sandwich Cookies        61   
6             0           1  Chocolate Sandwich Cookies        61   
7             0           1  Chocolate Sandwich Cookies        61   
8             0           1  Chocolate Sandwich Cookies        61   
9             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0             19     5.8          1987          1987   3139998      138  ...   
1             19     5.8          1989          1989   1977647      138  ...   
2             19     5.8         11433         11433    389851      709  ...   
3             19     5.8         12198         12198    652770      764  ...   
4             19     5.8         12200         12200   1813452      764  ...   
5             19     5.8         12372         12372   1701441      777  ...   
6             19     5.8         13096         13096   1871483      825  ...   
7             19     5.8         14603         14603   1290456      910  ...   
8             19     5.8         17065         17065    369558     1052  ...   
9             19     5.8         17070         17070    589712     1052  ...   

  order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                11                     3.0                  5          0   
1                17                    20.0                  1          1   
2                21                     6.0                 20          0   
3                13                    11.0                 10          0   
4                17                     9.0                 11          1   
5                 7                    26.0                  7          0   
6                14                    30.0                  2          0   
7                10                    30.0                  1          0   
8                20                    19.0                  1          0   
9                12                    15.0                  2          1   

   _merge  _merge2    price_range_loc   busiest_day  busiest_days  \
0    both     both  Mid-range product  Regular busy  Regular busy   
1    both     both  Mid-range product  Regular busy  Regular busy   
2    both     both  Mid-range product   Busiest day  Busiest days   
3    both     both  Mid-range product  Regular busy    Least busy   
4    both     both  Mid-range product    Least busy    Least busy   
5    both     both  Mid-range product  Regular busy  Busiest days   
6    both     both  Mid-range product  Regular busy  Regular busy   
7    both     both  Mid-range product  Regular busy    Least busy   
8    both     both  Mid-range product  Regular busy  Busiest days   
9    both     both  Mid-range product  Regular busy  Busiest days   

  busiest_period_of_day_loc  
0               Most orders  
1            Average orders  
2            Average orders  
3               Most orders  
4            Average orders  
5            Average orders  
6               Most orders  
7               Most orders  
8            Average orders  
9               Most orders  

[10 rows x 23 columns]

In [12]:
# Group data with pandas

df.groupby('product_name')

In [13]:
# Perform a single aggregation

df.groupby('department_id').agg({'number_of_order': ['mean']})

number_of_order
                         mean
department_id                
1                   15.577493
2                   17.320781
3                   16.084944
4                   17.530458
5                   14.763075
6                   16.658449
7                   17.031590
8                   15.076662
9                   15.447580
10                  18.681852
11                  15.447411
12                  14.327957
13                  16.548642
14                  16.960241
15                  16.121948
16                  17.803851
17                  15.593633
18                  19.674252
19                  16.899756
20                  16.255442
21                  25.535479

In [14]:
df.groupby('department_id')['number_of_order'].mean()

department_id
1     15.577493
2     17.320781
3     16.084944
4     17.530458
5     14.763075
6     16.658449
7     17.031590
8     15.076662
9     15.447580
10    18.681852
11    15.447411
12    14.327957
13    16.548642
14    16.960241
15    16.121948
16    17.803851
17    15.593633
18    19.674252
19    16.899756
20    16.255442
21    25.535479
Name: number_of_order, dtype: float64

In [15]:
df.groupby('department_id').number_of_order.mean()

department_id
1     15.577493
2     17.320781
3     16.084944
4     17.530458
5     14.763075
6     16.658449
7     17.031590
8     15.076662
9     15.447580
10    18.681852
11    15.447411
12    14.327957
13    16.548642
14    16.960241
15    16.121948
16    17.803851
17    15.593633
18    19.674252
19    16.899756
20    16.255442
21    25.535479
Name: number_of_order, dtype: float64

In [16]:
# Perform multiple aggregations

df.groupby('department_id').agg({'number_of_order': ['mean', 'min', 'max']})

number_of_order        
                         mean min max
department_id                        
1                   15.577493   1  99
2                   17.320781   1  96
3                   16.084944   1  99
4                   17.530458   1  99
5                   14.763075   1  99
6                   16.658449   1  99
7                   17.031590   1  99
8                   15.076662   1  98
9                   15.447580   1  99
10                  18.681852   1  99
11                  15.447411   1  99
12                  14.327957   1  99
13                  16.548642   1  99
14                  16.960241   1  99
15                  16.121948   1  99
16                  17.803851   1  99
17                  15.593633   1  99
18                  19.674252   1  99
19                  16.899756   1  99
20                  16.255442   1  99
21                  25.535479   1  98

In [17]:
# Find the aggregated mean for the entire dataframe

ords_prods_merge.groupby('department_id')['number_of_order'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: number_of_order, dtype: float64

## 3. Analyze the results and differences

Both the entire dataframe and the subset have 'department_id' from 1 to 21. But their means of 'number_of_order' by 'department_id' are different, since the subset only has 1 million rows while the entire dataframe has much more rows.

## 4. Create a loyalty flag using transform() and loc()

In [21]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['number_of_order'].transform(np.max)

C:\Users\admin\AppData\Local\Temp\ipykernel_20688\506111299.py:1: FutureWarning: The provided callable <function max at 0x000002B1B85C2C00> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['number_of_order'].transform(np.max)


In [22]:
ords_prods_merge.head(15)

Unnamed: 0_x  product_id                product_name  aisle_id  \
0              0           1  Chocolate Sandwich Cookies        61   
1              0           1  Chocolate Sandwich Cookies        61   
2              0           1  Chocolate Sandwich Cookies        61   
3              0           1  Chocolate Sandwich Cookies        61   
4              0           1  Chocolate Sandwich Cookies        61   
5              0           1  Chocolate Sandwich Cookies        61   
6              0           1  Chocolate Sandwich Cookies        61   
7              0           1  Chocolate Sandwich Cookies        61   
8              0           1  Chocolate Sandwich Cookies        61   
9              0           1  Chocolate Sandwich Cookies        61   
10             0           1  Chocolate Sandwich Cookies        61   
11             0           1  Chocolate Sandwich Cookies        61   
12             0           1  Chocolate Sandwich Cookies        61   
13             0           1  Chocolate Sandwich Cookies        61   
14             0           1  Chocolate Sandwich Cookies        61   

    department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0              19     5.8          1987          1987   3139998      138  ...   
1              19     5.8          1989          1989   1977647      138  ...   
2              19     5.8         11433         11433    389851      709  ...   
3              19     5.8         12198         12198    652770      764  ...   
4              19     5.8         12200         12200   1813452      764  ...   
5              19     5.8         12372         12372   1701441      777  ...   
6              19     5.8         13096         13096   1871483      825  ...   
7              19     5.8         14603         14603   1290456      910  ...   
8              19     5.8         17065         17065    369558     1052  ...   
9              19     5.8         17070         17070    589712     1052  ...   
10             19     5.8         22082         22082   1798115     1379  ...   
11             19     5.8         23713         23713    497193     1480  ...   
12             19     5.8         23715         23715   1727826     1480  ...   
13             19     5.8         23716         23716   2694698     1480  ...   
14             19     5.8         23937         23937   3171753     1494  ...   

   days_since_prior_order  add_to_cart_order  reordered  _merge  _merge2  \
0                     3.0                  5          0    both     both   
1                    20.0                  1          1    both     both   
2                     6.0                 20          0    both     both   
3                    11.0                 10          0    both     both   
4                     9.0                 11          1    both     both   
5                    26.0                  7          0    both     both   
6                    30.0                  2          0    both     both   
7                    30.0                  1          0    both     both   
8                    19.0                  1          0    both     both   
9                    15.0                  2          1    both     both   
10                   11.0                  1          0    both     both   
11                    7.0                  1          0    both     both   
12                    7.0                  3          1    both     both   
13                   12.0                  4          1    both     both   
14                    2.0                  8          0    both     both   

      price_range_loc   busiest_day  busiest_days busiest_period_of_day_loc  \
0   Mid-range product  Regular busy  Regular busy               Most orders   
1   Mid-range product  Regular busy  Regular busy            Average orders   
2   Mid-range product   Busiest day  Busiest days            Average orders   
3   Mid-range product  Regular busy    Least busy    

In [23]:
ords_prods_merge.head(100)

Unnamed: 0_x  product_id                product_name  aisle_id  \
0              0           1  Chocolate Sandwich Cookies        61   
1              0           1  Chocolate Sandwich Cookies        61   
2              0           1  Chocolate Sandwich Cookies        61   
3              0           1  Chocolate Sandwich Cookies        61   
4              0           1  Chocolate Sandwich Cookies        61   
..           ...         ...                         ...       ...   
95             0           1  Chocolate Sandwich Cookies        61   
96             0           1  Chocolate Sandwich Cookies        61   
97             0           1  Chocolate Sandwich Cookies        61   
98             0           1  Chocolate Sandwich Cookies        61   
99             0           1  Chocolate Sandwich Cookies        61   

    department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0              19     5.8          1987          1987   3139998      138  ...   
1              19     5.8          1989          1989   1977647      138  ...   
2              19     5.8         11433         11433    389851      709  ...   
3              19     5.8         12198         12198    652770      764  ...   
4              19     5.8         12200         12200   1813452      764  ...   
..            ...     ...           ...           ...       ...      ...  ...   
95             19     5.8        179480        179480    602103    10831  ...   
96             19     5.8        184011        184011     49629    11119  ...   
97             19     5.8        185002        185002    317888    11186  ...   
98             19     5.8        186194        186194    682486    11243  ...   
99             19     5.8        189258        189258   2861386    11448  ...   

   days_since_prior_order  add_to_cart_order  reordered  _merge  _merge2  \
0                     3.0                  5          0    both     both   
1                    20.0                  1          1    both     both   
2                     6.0                 20          0    both     both   
3                    11.0                 10          0    both     both   
4                     9.0                 11          1    both     both   
..                    ...                ...        ...     ...      ...   
95                   23.0                  5          0    both     both   
96                   30.0                  1          0    both     both   
97                    2.0                  8          0    both     both   
98                    0.0                  2          0    both     both   
99                    3.0                  6          0    both     both   

      price_range_loc   busiest_day  busiest_days busiest_period_of_day_loc  \
0   Mid-range product  Regular busy  Regular busy               Most orders   
1   Mid-range product  Regular busy  Regular busy            Average orders   
2   Mid-range product   Busiest day  Busiest days            Average orders   
3   Mid-range product  Regular busy    Least busy               Most orders   
4   Mid-range product    Least busy    Least busy            Average orders   
..                ...           ...           ...                       ...   
95  Mid-range product  Regular busy    Least busy               Most orders   
96  Mid-range product  Regular busy  Busiest days               Most orders   
97  Mid-range product  Regular busy  Regular busy               Most orders   
98  Mid-range product  Regular busy    Least busy               Most orders   
99  Mid-range product  Regular busy  Regular busy               Most orders   

   max_order  
0         32  
1         32  
2          5  
3          3  
4          3  
..       ...  
95        10  
96        23  
97        26  
98        43  
99        21  

[100 rows x 24 columns]

In [24]:
# Tell panda not to assign any options regarding the max number of rows to display

pd.options.display.max_rows = None

In [25]:
ords_prods_merge.head(50)

Unnamed: 0_x  product_id                product_name  aisle_id  \
0              0           1  Chocolate Sandwich Cookies        61   
1              0           1  Chocolate Sandwich Cookies        61   
2              0           1  Chocolate Sandwich Cookies        61   
3              0           1  Chocolate Sandwich Cookies        61   
4              0           1  Chocolate Sandwich Cookies        61   
5              0           1  Chocolate Sandwich Cookies        61   
6              0           1  Chocolate Sandwich Cookies        61   
7              0           1  Chocolate Sandwich Cookies        61   
8              0           1  Chocolate Sandwich Cookies        61   
9              0           1  Chocolate Sandwich Cookies        61   
10             0           1  Chocolate Sandwich Cookies        61   
11             0           1  Chocolate Sandwich Cookies        61   
12             0           1  Chocolate Sandwich Cookies        61   
13             0           1  Chocolate Sandwich Cookies        61   
14             0           1  Chocolate Sandwich Cookies        61   
15             0           1  Chocolate Sandwich Cookies        61   
16             0           1  Chocolate Sandwich Cookies        61   
17             0           1  Chocolate Sandwich Cookies        61   
18             0           1  Chocolate Sandwich Cookies        61   
19             0           1  Chocolate Sandwich Cookies        61   
20             0           1  Chocolate Sandwich Cookies        61   
21             0           1  Chocolate Sandwich Cookies        61   
22             0           1  Chocolate Sandwich Cookies        61   
23             0           1  Chocolate Sandwich Cookies        61   
24             0           1  Chocolate Sandwich Cookies        61   
25             0           1  Chocolate Sandwich Cookies        61   
26             0           1  Chocolate Sandwich Cookies        61   
27             0           1  Chocolate Sandwich Cookies        61   
28             0           1  Chocolate Sandwich Cookies        61   
29             0           1  Chocolate Sandwich Cookies        61   
30             0           1  Chocolate Sandwich Cookies        61   
31             0           1  Chocolate Sandwich Cookies        61   
32             0           1  Chocolate Sandwich Cookies        61   
33             0           1  Chocolate Sandwich Cookies        61   
34             0           1  Chocolate Sandwich Cookies        61   
35             0           1  Chocolate Sandwich Cookies        61   
36             0           1  Chocolate Sandwich Cookies        61   
37             0           1  Chocolate Sandwich Cookies        61   
38             0           1  Chocolate Sandwich Cookies        61   
39             0           1  Chocolate Sandwich Cookies        61   
40             0           1  Chocolate Sandwich Cookies        61   
41             0           1  Chocolate Sandwich Cookies        61   
42             0           1  Chocolate Sandwich Cookies        61   
43             0           1  Chocolate Sandwich Cookies        61   
44             0           1  Chocolate Sandwich Cookies        61   
45             0           1  Chocolate Sandwich Cookies        61   
46             0           1  Chocolate Sandwich Cookies        61   
47             0           1  Chocolate Sandwich Cookies        61   
48             0           1  Chocolate Sandwich Cookies        61   
49             0           1  Chocolate Sandwich Cookies        61   

    department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0              19     5.8          1987          1987   3139998      138  ...   
1              19     5.8          1989          1989   1977647      138  ...   
2              19     5.8         11433         11433    389851      709  ...   
3              19     5.8         12198         12198    652770      764  ...   
4              19     5.8   

In [26]:
# Deriving columns with loc()

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [27]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [28]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [29]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [30]:
ords_prods_merge[['user_id', 'loyalty_flag', 'number_of_order']].head(60)

user_id      loyalty_flag  number_of_order
0       138  Regular customer               28
1       138  Regular customer               30
2       709      New customer                2
3       764      New customer                1
4       764      New customer                3
5       777  Regular customer               16
6       825      New customer                3
7       910  Regular customer               12
8      1052  Regular customer               10
9      1052  Regular customer               15
10     1379      New customer                1
11     1480    Loyal customer               36
12     1480    Loyal customer               38
13     1480    Loyal customer               39
14     1494  Regular customer                6
15     1494  Regular customer                7
16     1494  Regular customer                9
17     1540  Regular customer                4
18     1540  Regular customer                5
19     1540  Regular customer                6
20     1540  Regular customer                7
21     1540  Regular customer                8
22     1540  Regular customer                9
23     1540  Regular customer               14
24     1540  Regular customer               15
25     1540  Regular customer               18
26     1540  Regular customer               19
27     1540  Regular customer               20
28     1540  Regular customer               21
29     1540  Regular customer               22
30     1540  Regular customer               24
31     1540  Regular customer               25
32     1540  Regular customer               27
33     1540  Regular customer               28
34     1598      New customer                1
35     1598      New customer                2
36     1598      New customer                3
37     1647      New customer                1
38     1647      New customer                4
39     1647      New customer                6
40     1647      New customer                8
41     1647      New customer                9
42     2833      New customer                1
43     2850  Regular customer                7
44     2857  Regular customer               24
45     3010  Regular customer                5
46     3029      New customer                1
47     3395      New customer                4
48     3904      New customer                1
49     4006    Loyal customer                2
50     4006    Loyal customer                3
51     4006    Loyal customer                5
52     4006    Loyal customer                9
53     4006    Loyal customer               12
54     4122  Regular customer               11
55     4141      New customer                2
56     4343      New customer                1
57     4553      New customer                3
58     4845  Regular customer                6
59     6866      New customer                1

## 5. Check the basic statistics of the product prices for each loyalty category

In [32]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

The mean of product prices for Loyal customer (10.37) is a little lower than the means of product prices for Regular customer (13.29) and New customer (12.5).

## 6. Create a spending flag for each user based on the average prices across all their orders

In [35]:
ords_prods_merge['average_prices'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\admin\AppData\Local\Temp\ipykernel_20688\1640056616.py:1: FutureWarning: The provided callable <function mean at 0x000002B1B85C3600> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_prices'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [36]:
ords_prods_merge.head(15)

Unnamed: 0_x  product_id                product_name  aisle_id  \
0              0           1  Chocolate Sandwich Cookies        61   
1              0           1  Chocolate Sandwich Cookies        61   
2              0           1  Chocolate Sandwich Cookies        61   
3              0           1  Chocolate Sandwich Cookies        61   
4              0           1  Chocolate Sandwich Cookies        61   
5              0           1  Chocolate Sandwich Cookies        61   
6              0           1  Chocolate Sandwich Cookies        61   
7              0           1  Chocolate Sandwich Cookies        61   
8              0           1  Chocolate Sandwich Cookies        61   
9              0           1  Chocolate Sandwich Cookies        61   
10             0           1  Chocolate Sandwich Cookies        61   
11             0           1  Chocolate Sandwich Cookies        61   
12             0           1  Chocolate Sandwich Cookies        61   
13             0           1  Chocolate Sandwich Cookies        61   
14             0           1  Chocolate Sandwich Cookies        61   

    department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0              19     5.8          1987          1987   3139998      138  ...   
1              19     5.8          1989          1989   1977647      138  ...   
2              19     5.8         11433         11433    389851      709  ...   
3              19     5.8         12198         12198    652770      764  ...   
4              19     5.8         12200         12200   1813452      764  ...   
5              19     5.8         12372         12372   1701441      777  ...   
6              19     5.8         13096         13096   1871483      825  ...   
7              19     5.8         14603         14603   1290456      910  ...   
8              19     5.8         17065         17065    369558     1052  ...   
9              19     5.8         17070         17070    589712     1052  ...   
10             19     5.8         22082         22082   1798115     1379  ...   
11             19     5.8         23713         23713    497193     1480  ...   
12             19     5.8         23715         23715   1727826     1480  ...   
13             19     5.8         23716         23716   2694698     1480  ...   
14             19     5.8         23937         23937   3171753     1494  ...   

   reordered  _merge  _merge2    price_range_loc   busiest_day  busiest_days  \
0          0    both     both  Mid-range product  Regular busy  Regular busy   
1          1    both     both  Mid-range product  Regular busy  Regular busy   
2          0    both     both  Mid-range product   Busiest day  Busiest days   
3          0    both     both  Mid-range product  Regular busy    Least busy   
4          1    both     both  Mid-range product    Least busy    Least busy   
5          0    both     both  Mid-range product  Regular busy  Busiest days   
6          0    both     both  Mid-range product  Regular busy  Regular busy   
7          0    both     both  Mid-range product  Regular busy    Least busy   
8          0    both     both  Mid-range product  Regular busy  Busiest days   
9          1    both     both  Mid-range product  Regular busy  Busiest days   
10         0    both     both  Mid-range product  Regular busy  Regular busy   
11         0    both     both  Mid-range product  Regular busy    Least busy   
12         1    both     both  Mid-range product    Least busy    Least busy   
13         1    both     both  Mid-range product  Regular busy  Regular busy   
14         0    both     both  Mid-range product  Regular busy    Least busy   

    busiest_period_of_day_loc max_order      loyalty_flag average_prices  
0                 Most orders        32  Regular customer       6.935811  
1              Average orders        32  Regular customer       6.935811  
2              Average orders         5      New customer       7.930208  
3    

In [37]:
# Create a spending flag for each user

ords_prods_merge.loc[ords_prods_merge['average_prices'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['average_prices'] >= 10, 'spending_flag'] = 'High spender'

In [38]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

## 7. Create an order frequency flag

In [40]:
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\admin\AppData\Local\Temp\ipykernel_20688\2588299015.py:1: FutureWarning: The provided callable <function median at 0x000002B1B873ADE0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [41]:
ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0             19     5.8          1987          1987   3139998      138  ...   
1             19     5.8          1989          1989   1977647      138  ...   
2             19     5.8         11433         11433    389851      709  ...   
3             19     5.8         12198         12198    652770      764  ...   
4             19     5.8         12200         12200   1813452      764  ...   

  _merge2    price_range_loc   busiest_day  busiest_days  \
0    both  Mid-range product  Regular busy  Regular busy   
1    both  Mid-range product  Regular busy  Regular busy   
2    both  Mid-range product   Busiest day  Busiest days   
3    both  Mid-range product  Regular busy    Least busy   
4    both  Mid-range product    Least busy    Least busy   

   busiest_period_of_day_loc  max_order      loyalty_flag average_prices  \
0                Most orders         32  Regular customer       6.935811   
1             Average orders         32  Regular customer       6.935811   
2             Average orders          5      New customer       7.930208   
3                Most orders          3      New customer       4.972414   
4             Average orders          3      New customer       4.972414   

  spending_flag median_days_since_prior_order  
0   Low spender                           8.0  
1   Low spender                           8.0  
2   Low spender                           8.0  
3   Low spender                          11.0  
4   Low spender                          11.0  

[5 rows x 28 columns]

In [42]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] > 10) & (ords_prods_merge['median_days_since_prior_order'] <= 20), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [43]:
ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  user_id  ...  \
0             19     5.8          1987          1987   3139998      138  ...   
1             19     5.8          1989          1989   1977647      138  ...   
2             19     5.8         11433         11433    389851      709  ...   
3             19     5.8         12198         12198    652770      764  ...   
4             19     5.8         12200         12200   1813452      764  ...   

     price_range_loc   busiest_day  busiest_days  busiest_period_of_day_loc  \
0  Mid-range product  Regular busy  Regular busy                Most orders   
1  Mid-range product  Regular busy  Regular busy             Average orders   
2  Mid-range product   Busiest day  Busiest days             Average orders   
3  Mid-range product  Regular busy    Least busy                Most orders   
4  Mid-range product    Least busy    Least busy             Average orders   

   max_order      loyalty_flag  average_prices spending_flag  \
0         32  Regular customer        6.935811   Low spender   
1         32  Regular customer        6.935811   Low spender   
2          5      New customer        7.930208   Low spender   
3          3      New customer        4.972414   Low spender   
4          3      New customer        4.972414   Low spender   

  median_days_since_prior_order order_frequency_flag  
0                           8.0    Frequent customer  
1                           8.0    Frequent customer  
2                           8.0    Frequent customer  
3                          11.0     Regular customer  
4                          11.0     Regular customer  

[5 rows x 29 columns]

## 8. Ensure the notebook is clean and structured and that the code is well commented

## 9. Export the dataframe as a pickle file in Prepared Data folder

In [46]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_new_cols_grouped.pkl'))